### PR TITLE
Closes #45 - Bootstrap env

### LINKED ISSUE #45: Feature request: bootstrap-env
The idea would be:
```console
$ vault bootstrap-env path1/ path2/ some_secret -- some command
```
Would read all secrets from path1, path2 and some_secret, put them in the environment, and then `exec`  `some command`.

A `--backup /some/location` would make it so that a yaml file with those secrets would be written locally at that location. If the vault is not available and the backup file is, then the backup file is used transparently and the command runs (though with a `logger.exception`).

Use case: start an app that reads settings from the environment (this would be the systemd entrypoint). The `exec` part is important so that we don't have to manage signals or what, the process is replaced with the real process as soon as possible.

The ``--backup`` flag would allow to use this way of starting the app even if your vault instance doesn't have 100% high availability yet (if the app restarts when your vault is down, your app can still run)

### PR DESCRIPTION
**Pull request by ewjoachim**
Closes #45

PR based on #46 , please merge that one first ! *EDIT* done ! thx

With this, one might be able to start loading secrets at process launch even if their vault doesn't offer high availablility yet.

Of course, saving secrets locally is sub-par, but any other solution would involve saving secrets locally one way or another, so at least this one is one step ahead in the future.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,8 @@
 # Config file for automatic testing at travis-ci.org
 language: python
 cache: pip
-
+sudo: required
+dist: xenial
```
Comment (ewjoachim) [inline .travis.yml]: This is required for python 3.7

**Comment by mgu** [comment]
Could you explicit the behavior with nested dicts ? (What will be the naming of the environment variables ?).
The "backup" feature looks like the opposite of what people want to achieve with HVault.

**Comment by ewjoachim** [comment]
> The "backup" feature looks like the opposite of what people want to achieve with HVault.

Yes, that's why I said:

> Of course, saving secrets locally is sub-par, but any other solution would involve saving secrets locally one way or another, so at least this one is one step ahead in the future.

Let's discuss that in the ticket !

> Could you explicit the behavior with nested dicts ? (What will be the naming of the environment variables ?).

Yes, my initial stance would be that this would only make sense with string secrets, so any value is simply "strigized". But I was hesitation a lot with yamlize or jsonify.

string:
- should suffice 99% of the time, so "yagni", we're still <v1, let's wait for real usecases

json:
- will fit in a single line
- is yaml compatible

yaml:
- is consistent

Having no idea how to choose, I applied the Ockam razor and chose the simplest solution.

**Comment by mgu** [comment]
let me give you an example :)
```sh
% vault --config-file ./config.yaml set myproject/SECRET_KEY  xxx
Done
% vault --config-file ./config.yaml set myproject/postgresql/default pgsql://dsn
Done
% vault --config-file ./config.yaml set myproject/infra/rmq/default rmq://dsn
% vault --config-file ./config.yaml get-all myproject/
---
myproject/:
  SECRET_KEY: xxx
  infra:
    rmq:
      default: rmq://dsn
  postgresql:
    default: pgsql://dsn

% vault --config-file ./config.yaml bootstrap-env --path myproject env
...
SECRET_KEY=xxx
infra={'rmq': {'default': 'rmq://dsn'}}
postgresql={'default': 'pgsql://dsn'}
```

I don't like these dicts and I think we could _flatten_ the tree and have something like : 
```
SECRET_KEY=xxx
infra_rmq_defaul=rmq://dsn'
postgresql_default=pgsql://dsn'
```
(we could use another separator)

**Comment by ewjoachim** [comment]
That’s an interesting approach... Would it be ok to do it as a separate PR and build incrementally ? I’m affraid we might run into more complexity, for example, what happens when there’s a list (should we expose it as a dictionnary with integer keys ? Shall we expose the list length as an additional element ? What about booleans ?

**Comment by mgu** [comment]
was it a good idea to support non-string values ? :D

**Comment by ewjoachim** [comment]
In the general case, I’d say yes but... :joy:

Is it ok if we just don’t support strings for this command for now ? Not crashing, this would be dangerous, but just putting a blank string is the object is not a string ?

**Comment by mgu** [comment]
really, I don't care about the value and its format, only about being able to use keys

**Comment by ewjoachim** [comment]
You never answered
> Would it be ok to do it as a separate PR and build incrementally ?

I guess it's a no.


After testing it, there is also a problem when we want to recursively expose a folder 2 levels deep
```console 
$ vault set kapoue/youpi/a b
Done
$ vault set kapoue/youpi/c d
Done
$ vault bootstrap-env -p kapoue/youpi env |grep youpi
youpi={'a': 'b', 'c': 'd'}
```

**Comment by mgu** [comment]
> Would it be ok to do it as a separate PR and build incrementally ?

I would prefer a first attempt in this PR :)

**Comment by ewjoachim** [comment]
> I would prefer a first attempt in this PR :)
Seen with mgu: we'll merge now without it but do a PR right away to implement it.

**Pull request merged**